multirpc/subpub/discovery.go

```diff
@@ -3,13 +3,100 @@ package subpub
+	discovery "github.com/libp2p/go-libp2p-discovery"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	dhtopts "github.com/libp2p/go-libp2p-kad-dht/opts"
 	"go.vocdoni.io/dvote/log"
 )
 
+// setupDiscovery creates a DHT discovery service and attaches it to the libp2p Host.
+// This lets us automatically discover peers and connect to them.
+func (ps *SubPub) setupDiscovery(ctx context.Context) {
+	var err error
+
+	// Set a function as stream handler. This function is called when a peer
+	// initiates a connection and starts a stream with this peer.
+	ps.Host.SetStreamHandler(protocol.ID(ps.Topic), ps.handleStream)
+
+	// Start a DHT, for use in peer discovery. We can't just make a new DHT
+	// client because we want each peer to maintain its own local copy of the
+	// DHT, so that the bootstrapping node of the DHT can go down without
+	// inhibiting future peer discovery.
+
+	// Let's try to apply some tunning for reducing the DHT fingerprint
+	opts := []dhtopts.Option{
+		dht.RoutingTableLatencyTolerance(time.Second * 20),
+		dht.BucketSize(20),
+		dht.MaxRecordAge(1 * time.Hour),
+	}
+
+	// Note that we don't use ctx here, since we stop via the Close method.
+	ps.dht, err = dht.New(context.Background(), ps.Host, opts...)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if !ps.NoBootStrap {
+		// Bootstrap the DHT. In the default configuration, this spawns a Background
+		// thread that will refresh the peer table every five minutes.
+		log.Info("bootstrapping the DHT")
+		// Note that we don't use ctx here, since we stop via the Close method.
+		if err := ps.dht.Bootstrap(context.Background()); err != nil {
+			log.Fatal(err)
+		}
+
+		// Let's connect to the bootstrap nodes first. They will tell us about the
+		// other nodes in the network.
+		bootnodes := dht.DefaultBootstrapPeers
+		if len(ps.BootNodes) > 0 {
+			bootnodes = parseMultiaddress(ps.BootNodes)
+		}
+
+		log.Info("connecting to bootstrap nodes...")
+		log.Debugf("bootnodes: %+v", bootnodes)
+		var wg sync.WaitGroup
+		for _, peerAddr := range bootnodes {
+			if peerAddr == nil {
+				continue
+			}
+			peerinfo, err := libpeer.AddrInfoFromP2pAddr(peerAddr)
+			if err != nil {
+				log.Fatal(err)
+			}
+			if peerinfo == nil {
+				continue // nothing to do
+			}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				log.Debugf("trying %s", *peerinfo)
+				if err := ps.Host.Connect(ctx, *peerinfo); err != nil {
+					log.Debug(err)
+				} else {
+					log.Infof("connection established with bootstrap node: %s", peerinfo)
+				}
+			}()
+		}
+		wg.Wait()
+	}
+
+	// We use a rendezvous point "meet me here" to announce our location.
+	// This is like telling your friends to meet you at the Eiffel Tower.
+	ps.routing = discovery.NewRoutingDiscovery(ps.dht)
+
+	//// TODO: move here as well: (instead of in Subscribe())
+	//  for { ps.discover(pctx) ; time.Sleep(ps.DiscoveryPeriod) }
+	//  go ps.advertise(ctx, ps.Topic)
+	//  go ps.advertise(ctx, ps.PubKey)
```

Comment:
why is this needed, advertising oneself (in addition to ps.Topic) on topic ps.PubKey? i couldn't find anywhere in the code where this is used.

Reply:
That's legacy. The protocol use to support contacting someone by its pubKey too. So (for instance) a bootstrap server could announce its pubKey in order to let clients find him and retrieve the list of gateways to connect with.

We have not used it yet but it might be useful in the future. However I can accept removing this from now, no problem.